src/openclon12/resource_tasks.cpp

```diff
@@ -2532,8 +2532,18 @@ class MapCopyTask : public MapTask
         Args.m_appDesc.m_Height = args.Height;
         Args.m_appDesc.m_usage = D3D12TranslationLayer::RESOURCE_USAGE_STAGING;
         Args.m_appDesc.m_bindFlags = D3D12TranslationLayer::RESOURCE_BIND_NONE;
-        Args.m_appDesc.m_cpuAcess = D3D12TranslationLayer::RESOURCE_CPU_ACCESS_READ | D3D12TranslationLayer::RESOURCE_CPU_ACCESS_WRITE;
-        Args.m_heapDesc = CD3DX12_HEAP_DESC(0, D3D12_HEAP_TYPE_READBACK);
+
+        if (flags | CL_MAP_WRITE)
+        {
+            Args.m_appDesc.m_cpuAcess = D3D12TranslationLayer::RESOURCE_CPU_ACCESS_WRITE | D3D12TranslationLayer::RESOURCE_CPU_ACCESS_READ;
+            Args.m_heapDesc = CD3DX12_HEAP_DESC(0, D3D12_CPU_PAGE_PROPERTY_WRITE_COMBINE, D3D12_MEMORY_POOL_L0);

```

Comment:
This will destroy performance on GPUs which use the CPU MMU for accessing the data, which includes the software rasterizer, WARP. It'll force them to read from write-combine memory. That's why this code typically uses the abstractions of `UPLOAD`, `READBACK`, and `DEFAULT`, to handle GPUs with VRAM and without, as well as cache-coherent UMA where `UPLOAD` should be `WRITE_BACK`.

Reply:
in previous code, validation layer will report error if trying to transit a READBACK resource to COPY_SOURCE state if just using READ_BACK for everything.
We need to copy DEFAULT resource to mappable resource on Record(), we can't use UPLOAD heap for mappable resource since it can't be transit to COPY_DEST.
A CUSTOM heap with WRITEBACK/WRITECOMBINE can solve the problem. But there are performance tradeoff:
In CPU write-only scenario, a WRITECOMBINE is most optimized.
In CPU writable and readable scenario, WRITEBACK is more optimized for CPU read, CPU write performance will be mid;  WRITECOMBINE is more optimized for CPU write, but CPU Read performance will be destroyed
Am I understanding this correctly?
Maybe we can go:
WRITECOMBINE for CL_MAP_WRITE-only since there won't be CPU read in this scenario
WRITEBACK for (CL_MAP_WRITE | CL_MAP_READ)
Is this seems viable to you?

Reply:
Apps should never explicitly ask for write-combine, because some drivers (like WARP) read using the same CPU page mappings, and that's problematic. Instead, `GetCustomHeapProperties(D3D12_HEAP_TYPE_UPLOAD)` returns an appropriate heap properties struct for write-only data.

I don't recall offhand if this `Args.m_heapDesc` gets translated to a custom heap type using `GetCustomHeapProperties` after this point. If so, then using `READBACK`/`UPLOAD` here should be fine. If it doesn't, then instead of explicit write-combine, please use `GetCustomHeapProperties`.

Reply:
ACK. I updated this PR, pls take a look.
and GetCustomHeapProperties returns a TYPE_CUSTOM heap.